### PR TITLE
Merge back changes to run CE under Windows Subsystem for Linux

### PR DESCRIPTION
**Pull request by AndrewPardoe**
This small set of changes makes it possible to run the CE under the Windows Subsystem for Linux. Most compilers are run under Linux, as usual, but MSVC is run under Windows rather than Wine. 

A few of the configuration details are hard-coded:
- The current state assumes compilers & Windows SDKs are located at c:\str
- The compilers are assumed to be just an unzipping of the [daily MSVC NuGet packages](https://aka.ms/dailyMSVC)
- The current state assumes there is a temporary directory at c:\tmp

Because the code isn't generalize you may need to look at the code or contact me for details on setup.

The most impactful change is moving the existing Wine support to a Wine-specific configuration file and creating a WSL-specific configuration file.

### CODE REVIEW

File: lib/compilers/WSL-CL.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2012-2017, Matt Godbolt
```
Comment (RabsRincon) [inline lib/compilers/WSL-CL.js]: You can probably this to your name if you're comfortable with it

File: lib/compilers/WSL-CL.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            '/FAsc',
+            '/c',
+            '/Fa' + this.filename(outputFilename),
+            '/Fo' + this.filename(outputFilename + '.obj')
+        ];
+    };
+    return compile.initialise();
+}
+
+module.exports = compileCl;
```
Comment (RabsRincon) [inline lib/compilers/WSL-CL.js]: Missing EOF

File: etc/config/c++.defaults.properties [hunk 3]
```diff
@@ -12,8 +13,24 @@ binaryHideFuncRe=^(__.*|_(init|start|fini)|(de)?register_tm_clones|call_gmon_sta
...
+#compiler.cl19_64.name=x86-64 CL 19 2017 RTW
+#compiler.cl19_32.exe=/opt/compiler-explorer/windows/19.10.25017/lib/native/bin/amd64_x86/cl.exe
+#compiler.cl19_32.name=x86 CL 19 2017 RTW
+#compiler.cl19_arm.exe=/opt/compiler-explorer/windows/19.10.25017/lib/native/bin/amd64_arm/cl.exe
+#compiler.cl19_arm.name=ARM CL 2017 RTW
 
-#################################
-#################################
-# Installed libs (See c++.amazon.properties for a scheme of libs group)
-libs=
```
Comment (RabsRincon) [inline etc/config/c++.defaults.properties]: This deleted lines should probably stay. Local users might use this as a template for their .local and a reminder that you can add your own libs is never bad

Comment (AndrewPardoe) [reply etc/config/c++.defaults.properties]: That was a mistake, thank you for noticing!

File: lib/compilers/WSL-CL.js (hunk 1 above)
Comment (AndrewPardoe) [reply lib/compilers/WSL-CL.js]: Yes, Matt told me this morning to do that but I missed it when merging in his changes. Thank you!

File: lib/compilers/WSL-CL.js (hunk 2 above)
Comment (AndrewPardoe) [reply lib/compilers/WSL-CL.js]: Fixed.

**Comment by brinkrob** [comment]
I spent some time today trying to get Andrew's branch to work on my machine.  First off, thanks for the work.  I got it working with just a few hiccups.

1) The "c:\tmp" folder when run under WSL is required to exist.  I was getting weird errors at startup till I figured out that I needed to create it manually.  It would be nice if that folder got created automatically if needed.

**Comment by AndrewPardoe** [comment]
It's not just the c:\tmp path (which can be specialized on startup), it's also the location of the compiler in c:\str (which is not ideal.) 

The first limitation stems from WSL's inability to read Windows environment variables. There are a number of similar issues if you [query on the Bash on Windows issues list](https://github.com/Microsoft/BashOnWindows/issues?q=is%3Aopen+is%3Aissue+label%3Ainterop). 

Referencing @benhillis who won't be able to fix this immediately :)

**Comment by mattgodbolt** [comment]
Thanks both @brinkrob and @AndrewPardoe for this. I'm going to merge, warts'n'all, and we'll improve as we can once/if WSL has the features necessary.

**Pull request merged**